File: app/js/flows.js
/**
 * Flow functions
 *
 * - Identity
 * - Planar
 * - Radial
 */
IdentityFlow = function() {};

IdentityFlow.prototype.transform = function(data) {
  return data;
};

PlanarFlow = function(_w, _u, _b) {
  this.w0 = _w[0];
  this.w1 = _w[1];
  this.u0 = _u[0];
  this.u1 = _u[1];
  this.b = _b;
};
PlanarFlow.prototype.transform = function(z) {
  var flow = this;

  flow.w = [flow.w0, flow.w1];
  flow.u = [flow.u0, flow.u1];

  return z.map(function(d) {
    return math.add(d, math.multiply(flow.u, math.tanh(math.add(math.dot(flow.w, d), flow.b))));
  })
};
PlanarFlow.prototype.updateParams = function(newParams) {
  for (var param in newParams) {
    if (newParams.hasOwnProperty(param)) {
      this[param] = newParams[param];
    }
  }
};

File: normflows/normflows/flows.py
"""Implementations of planar and radial flows.
"""
from typing import Union

from autograd import numpy as np


def planar_flow(z: np.ndarray,
                w: np.ndarray,
                u: np.ndarray,
                b: Union[int, float],
                h=np.tanh) -> np.ndarray:
    """Apply a planar flow to each element of `samples`

    :param z: numpy array, samples to be transformed
        Shape: (n_samples, n_dim)
    :param u: numpy array, parameter of flow
    :param w: numpy array, parameter of flow
    :param b: numeric, parameter of flow
    :param h: callable, non-linear function (default tanh)
    :returns: numpy array, transformed samples

    Transforms given samples according to the planar flow
    :math:`f(z) = z + uh(w^Tz + b)`
    """
    u = _get_uhat(u, w)
    assert np.dot(u, w) >= -1, f'Flow is not guaranteed to be invertible (u^Tw < -1: {w._value, u._value})'
    assert np.dot(u, w) >= -1, f'Flow is not guaranteed to be invertible (u^Tw < -1: {w._value, u._value})'
    return z + np.outer(np.tanh(np.matmul(z, w) + b), u)


def _get_uhat(u, w):
    return u + (m(np.dot(w, u)) - np.dot(w, u)) * (w / (np.linalg.norm(w) ** 2))


def m(x):
    return -1 + np.log(1 + np.exp(x))
